apps/prairielearn/src/models/course-instances.ts

```diff
@@ -0,0 +1,52 @@
+import { callValidatedOneRow } from '@prairielearn/postgres';
+import { z } from 'zod';
+import { IdSchema } from '../lib/db-types';
+import { idsEqual } from '../lib/id';
+
+const CourseInstanceAuthzSchema = z.object({
+  course_instances: z.array(
+    z.object({
+      short_name: z.string().nullable(),
+      long_name: z.string().nullable(),
+      id: IdSchema,
+      formatted_start_date: z.string(),
+      formatted_end_date: z.string(),
+      has_course_instance_permission_view: z.boolean(),
+    }),
+  ),
+});
+
+export async function selectAuthorizedCourseInstancesForCourse({
+  course_id,
+  user_id,
+  authn_user_id,
+  is_administrator,
+}: {
+  course_id: string;
+  user_id: string;
+  authn_user_id: string;
+  is_administrator: boolean;
+}) {
+  const { course_instances: authnCourseInstances } = await callValidatedOneRow(
+    'course_instances_with_staff_access',
+    [authn_user_id, is_administrator, course_id],
```

Comment:
The value of `is_administrator` here is for the effective user, not the authenticated user. As a result, if the authenticated user is an administrator without explicit staff access, this list will return empty, since `is_administrator` is false. This should be updated to use the `is_administrator` value of the authenticated user (I believe this is `res.locals.access_as_administrator`), not the effective user.

Reply:
A very good point; I think `res.locals.authn_is_administrator` is what we actually want. `res.locals.access_as_administrator` just reflects whether or not the `pl_access_as_administrator` cookie is set to `active`.

Reply:
I added `authn_is_administrator` in 7e21f22. Can you check if that looks correct now?

Reply:
The missing course instance issue is now resolved. Other tests seem to mainly work, though I haven't tested exhaustively (about to leave).